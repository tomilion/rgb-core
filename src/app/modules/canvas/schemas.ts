export interface CreateCanvasPayload {
    canvasId: number;
    costPerPixel: bigint;
    startBlockHeight: bigint;
    endBlockHeight: bigint;
    width: number;
    height: number;
    timeBetweenDraws: number;
    colourPalette: Buffer;
    maxPixelsPerTransaction: number;
}

export interface ChangeCanvasPayload {
    canvasId: number;
    costPerPixel?: bigint | null;
    startBlockHeight?: bigint | null;
    endBlockHeight?: bigint | null;
    width?: number | null;
    height?: number | null;
    timeBetweenDraws?: number | null;
    colourPalette?: Buffer | null;
    maxPixelsPerTransaction?: number | null;
}

export enum CanvasState {
    PENDING = 0,
    ACTIVE = 1,
    COMPLETE = 2,
}

export interface CanvasPayload {
    ownerId: Buffer;
    costPerPixel: bigint;
    startBlockHeight: bigint;
    endBlockHeight: bigint;
    width: number;
    height: number;
    timeBetweenDraws: number;
    colourPalette: Buffer;
    maxPixelsPerTransaction: number;
    state: number;
}

export const canvasSchema = {
    $id: "canvas/canvas",
    type: "object",
    required: ["ownerId", "costPerPixel", "startBlockHeight", "endBlockHeight", "width", "height", "timeBetweenDraws", "colourPalette", "maxPixelsPerTransaction"],
    properties: {
        ownerId: { fieldNumber: 1, dataType: "bytes", minLength: 20, maxLength: 20 },
        costPerPixel: { fieldNumber: 2, dataType: "uint64" },
        startBlockHeight: { fieldNumber: 3, dataType: "uint64" },
        endBlockHeight: { fieldNumber: 4, dataType: "uint64" },
        width: { fieldNumber: 5, dataType: "uint32" },
        height: { fieldNumber: 6, dataType: "uint32" },
        timeBetweenDraws: { fieldNumber: 7, dataType: "uint32" },
        colourPalette: { fieldNumber: 8, dataType: "bytes", minLength: 48, maxLength: 48 },
        maxPixelsPerTransaction: { fieldNumber: 9, dataType: "uint32" },
        state: { fieldNumber: 10, dataType: "uint32", default: CanvasState.PENDING },
    },
};

export interface AccountPayload {
    lastBlockHeight: number;
}

export const accountSchema = {
    $id: "canvas/account",
    type: "object",
    required: ["lastBlockHeight"],
    properties: {
        lastBlockHeight: { fieldNumber: 1, dataType: "uint32" },
    },
};

export interface PendingPayload {
    canvasIds: number[];
}

export const pendingSchema = {
    $id: "canvas/pending",
    type: "object",
    required: ["canvasIds"],
    properties: {
        canvasIds: { fieldNumber: 1, type: "array", items: { dataType: 'uint32' } },
    },
};

export interface ActivePayload {
    canvasIds: number[];
}

export const activeSchema = {
    $id: "canvas/active",
    type: "object",
    required: ["canvasIds"],
    properties: {
        canvasIds: { fieldNumber: 1, type: "array", items: { dataType: 'uint32' } },
    },
};

export interface CompletePayload {
    canvasIds: number[];
}

export const completeSchema = {
    $id: "canvas/complete",
    type: "object",
    required: ["canvasIds"],
    properties: {
        canvasIds: { fieldNumber: 1, type: "array", items: { dataType: 'uint32' } },
    },
};

export interface AddressPayload {
    address: Buffer;
}

export const addressSchema = {
    $id: "canvas/address",
    type: "object",
    required: ["address"],
    properties: {
        address: { fieldNumber: 1, dataType: "bytes", minLength: 20, maxLength: 20 },
    },
};

export interface DrawPixelPayload {
    canvasId: number;
    coords: Buffer;
    colours: Buffer;
}

export interface DrawPixelPayloadJSON {
    canvasId: number;
    coords: string;
    colours: string;
}

export const drawPixelSchema = {
    $id: "canvas/drawPixel/asset",
    title: "DrawPixelAsset transaction asset for Canvas module",
    type: "object",
    required: ["canvasId", "coords", "colours"],
    properties: {
        canvasId: { fieldNumber: 1, dataType: "uint32" },
        coords: { fieldNumber: 2, dataType: "bytes", minLength: 3, maxLength: 30000 },
        colours: { fieldNumber: 3, dataType: "bytes", minLength: 1, maxLength: 5000 },
    },
};

export interface CanvasAccount {
    canvas: { accountType: string }
}

export enum AccountType {
    Admin = "ADMIN",
    Wallet = "WALLET",
    Default = "DEFAULT",
}

export interface CanvasResponse {
    ownerId: string;
    costPerPixel: number;
    startBlockHeight: number;
    endBlockHeight: number;
    width: number;
    height: number;
    timeBetweenDraws: number;
    colourPalette: string;
    maxPixelsPerTransaction: number;
    state: number;
}

export type CanvasId = {
    canvasId: number;
};

export type PixelChangeSubmitted = {
    address: string;
    transactionId: string;
    pixel: DrawPixelPayload;
};

export type PixelChangeCommitted = {
    address: string;
    transactionId: string;
    blockHeight: number;
    pixel: DrawPixelPayload;
};
