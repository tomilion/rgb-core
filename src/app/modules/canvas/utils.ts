export const serialiseCanvasId = (canvasId: number) => `canvas-${canvasId}`;

export const serialisePixelId = (canvasId: number, xCoord: number, yCoord: number) =>
    `${serialiseCanvasId(canvasId)}-pixel-${xCoord}-${yCoord}`;

export const serialiseAccountId = (canvasId: number, address: string) => `${serialiseCanvasId(canvasId)}-account-${address}`;
