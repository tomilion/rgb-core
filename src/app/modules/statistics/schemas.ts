export interface TotalTransactionsPayload {
    totalTransactions: number;
}

export const totalTransactionsSchema = {
    $id: "statistics/totalTransactions",
    type: "object",
    required: ["totalTransactions"],
    properties: {
        totalTransactions: { fieldNumber: 1, dataType: "uint32" },
    },
};
