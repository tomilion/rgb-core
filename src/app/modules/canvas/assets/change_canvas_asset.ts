import { ApplyAssetContext, BaseAsset, codec, ValidateAssetContext } from "lisk-sdk";
import { serialiseCanvasId } from "../utils";
import { CanvasPayload, canvasSchema, CanvasState, ChangeCanvasPayload } from "../schemas";

export class ChangeCanvasAsset extends BaseAsset<ChangeCanvasPayload> {
    public static readonly ASSET_ID = 1;

    public name = "changeCanvas";
    public id = ChangeCanvasAsset.ASSET_ID;
    public schema = {
        $id: "canvas/changeCanvas-asset",
        title: "ChangeCanvasAsset transaction asset for canvas module",
        type: "object",
        required: ["canvasId"],
        properties: {
            canvasId: { fieldNumber: 1, dataType: "uint32" },
            costPerPixel: { fieldNumber: 2, dataType: "uint64" },
            startBlockHeight: { fieldNumber: 3, dataType: "uint64" },
            endBlockHeight: { fieldNumber: 4, dataType: "uint64" },
            width: { fieldNumber: 5, dataType: "uint32" },
            height: { fieldNumber: 6, dataType: "uint32" },
            timeBetweenDraws: { fieldNumber: 7, dataType: "uint32" },
            colourPalette: { fieldNumber: 8, dataType: "bytes", minLength: 48, maxLength: 48 },
            maxPixelsPerTransaction: { fieldNumber: 9, dataType: "uint32" },
            label: { fieldNumber: 10, dataType: "string", minLength: 0, maxLength: 64 },
        },
    };

    public validate(context: ValidateAssetContext<ChangeCanvasPayload>): void {
        const { asset, header } = context;

        if (asset.width !== undefined && asset.width !== null && (asset.width < 1 || asset.width > 10000))
        {
            throw new Error("Width invalid");
        }

        if (asset.height !== undefined && asset.height !== null && (asset.height < 1 || asset.height > 10000))
        {
            throw new Error("Height invalid");
        }

        if (asset.costPerPixel !== undefined && asset.costPerPixel !== null && asset.costPerPixel < 0)
        {
            throw new Error("Cost per pixel invalid");
        }

        // Checking undefined/null inline because storing condition in variable is too complex for linter ...
        if (asset.startBlockHeight !== undefined && asset.startBlockHeight !== null && asset.startBlockHeight < header.height)
        {
            throw new Error("Start block height cannot be in the past");
        }

        // Checking undefined/null inline because storing condition in variable is too complex for linter ...
        if (asset.endBlockHeight !== undefined && asset.endBlockHeight !== null && asset.endBlockHeight < header.height)
        {
            throw new Error("End block height cannot be in the past");
        }

        // Checking undefined/null inline because storing condition in variable is too complex for linter ...
        if (asset.startBlockHeight !== undefined &&
            asset.startBlockHeight !== null &&
            asset.endBlockHeight !== undefined &&
            asset.endBlockHeight !== null &&
            asset.endBlockHeight < asset.startBlockHeight)
        {
            throw new Error("End block height must be greater than start block height");
        }

        if (asset.maxPixelsPerTransaction !== undefined && asset.maxPixelsPerTransaction !== null && (asset.maxPixelsPerTransaction < 1 || asset.maxPixelsPerTransaction > 10000))
        {
            throw new Error("Max pixels per transaction invalid");
        }

        if (asset.colourPalette !== undefined && asset.colourPalette !== null && asset.colourPalette.length !== 48)
        {
            throw new Error("Colour palette invalid");
        }
    }

    public async apply(context: ApplyAssetContext<ChangeCanvasPayload>): Promise<void> {
        const { asset, stateStore, transaction } = context;
        const canvasId = serialiseCanvasId(asset.canvasId);
        const currentCanvas = await stateStore.chain.get(canvasId);

        if (currentCanvas === undefined)
        {
            throw new Error("Canvas does not exist");
        }

        const canvas = codec.decode<CanvasPayload>(canvasSchema, currentCanvas);

        if (canvas.ownerId.toString("hex") !== transaction.senderAddress.toString("hex"))
        {
            throw new Error("User invalid");
        }

        if (canvas.state === CanvasState.COMPLETE)
        {
            throw new Error("Canvas already completed");
        }

        if (canvas.state === CanvasState.ACTIVE)
        {
            if (asset.startBlockHeight !== null && asset.startBlockHeight !== undefined)
            {
                throw new Error("Cannot modify start block height on active canvas");
            }

            if (asset.endBlockHeight !== null && asset.endBlockHeight !== undefined)
            {
                throw new Error("Cannot modify end block height on active canvas");
            }
        }

        canvas.width = asset.width ?? canvas.width;
        canvas.height = asset.height ?? canvas.height;
        canvas.costPerPixel = asset.costPerPixel ?? canvas.costPerPixel;
        canvas.startBlockHeight = asset.startBlockHeight ?? canvas.startBlockHeight;
        canvas.endBlockHeight = asset.endBlockHeight ?? canvas.endBlockHeight;
        canvas.timeBetweenDraws = asset.timeBetweenDraws ?? canvas.timeBetweenDraws;
        canvas.colourPalette = asset.colourPalette ?? canvas.colourPalette;
        canvas.maxPixelsPerTransaction = asset.maxPixelsPerTransaction ?? canvas.maxPixelsPerTransaction;
        canvas.label = asset.label ?? canvas.label;

        if (canvas.startBlockHeight > canvas.endBlockHeight)
        {
            throw new Error("End block height must be greater than start block height");
        }

        await stateStore.chain.set(canvasId, codec.encode(canvasSchema, canvas));
    }
}
