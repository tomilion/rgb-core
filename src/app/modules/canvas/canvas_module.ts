import { AfterBlockApplyContext, AfterGenesisBlockApplyContext, BaseModule, BeforeBlockApplyContext, codec, TransactionApplyContext } from "lisk-sdk";
import { strict as assert } from "assert";
import { ChangeCanvasAsset } from "./assets/change_canvas_asset";
import { CreateCanvasAsset } from "./assets/create_canvas_asset";
import { DrawPixelAsset, DrawPixelPayload, drawPixelSchema } from "./assets/draw_pixel_asset";
import { serialiseCanvasId } from "./utils";
import { AccountType, ActivePayload, activeSchema, AddressPayload, addressSchema, CanvasAccount, CanvasPayload, CanvasResponse, canvasSchema, CanvasState, CompletePayload, completeSchema, PendingPayload, pendingSchema } from "./schemas";

export class CanvasModule extends BaseModule {
    public static readonly MODULE_ID = 1000;

    public name = "canvas";
    public id = CanvasModule.MODULE_ID;
    public accountSchema = {
        type: "object",
        properties: {
            accountType: {
                fieldNumber: 1,
                dataType: "string",
                maxLength: 64,
            },
        },
        default: {
            accountType: AccountType.Default.toString(),
        },
    };
    public actions = {
        getCanvas: async (params) => this.getCanvas(params),
        getPendingCanvases: async () => this.getPendingCanvases(),
        getActiveCanvases: async () => this.getActiveCanvases(),
        getCompleteCanvases: async () => this.getCompleteCanvases(),
    };
    public reducers = {
        getWalletAddress: async () => this.getWalletAddress(),
        getAdminAddress: async () => this.getAdminAddress(),
    };
    public transactionAssets = [
        new CreateCanvasAsset(),
        new ChangeCanvasAsset(),
        new DrawPixelAsset(),
    ];
    public events = ["started", "completed", "pixelChangeSubmitted", "pixelChangeCommitted"];

    // Lifecycle hooks
    public async beforeBlockApply(_input: BeforeBlockApplyContext): Promise<void> {
        // Get any data from stateStore using block info, below is an example getting a generator
        // const generatorAddress = getAddressFromPublicKey(_input.block.header.generatorPublicKey);
        // const generator = await _input.stateStore.account.get<TokenAccount>(generatorAddress);
    }

    public async afterBlockApply(context: AfterBlockApplyContext): Promise<void> {
        const { block, stateStore } = context;
        const pendingToCommit: number[] = [];
        const activeToCommit: number[] = [];

        const pendingBuffer = await stateStore.chain.get("canvas:pending");
        const pending = (pendingBuffer !== undefined) ? codec.decode<PendingPayload>(pendingSchema, pendingBuffer) : { canvasIds: [] };

        const activeBuffer = await stateStore.chain.get("canvas:active");
        const active = (activeBuffer !== undefined) ? codec.decode<ActivePayload>(activeSchema, activeBuffer) : { canvasIds: [] };

        const completeBuffer = await stateStore.chain.get("canvas:complete");
        const complete = (completeBuffer !== undefined) ? codec.decode<CompletePayload>(completeSchema, completeBuffer) : { canvasIds: [] };

        for (const canvasId of pending.canvasIds)
        {
            const canvasBuffer = await stateStore.chain.get(serialiseCanvasId(canvasId));
            assert(canvasBuffer !== undefined);
            const canvas = codec.decode<CanvasPayload>(canvasSchema, canvasBuffer);
            assert(canvas.state === CanvasState.PENDING);

            // Automatically starting canvas after passing start time
            if (canvas.startBlockHeight <= block.header.height)
            {
                canvas.state = CanvasState.ACTIVE;
                await stateStore.chain.set(serialiseCanvasId(canvasId), codec.encode(canvasSchema, canvas));
                activeToCommit.push(canvasId);
                this._channel.publish("canvas:started", { canvasId });
                continue;
            }

            pendingToCommit.push(canvasId);
        }

        for (const transaction of block.payload)
        {
            if (transaction.moduleID === CanvasModule.MODULE_ID &&
                transaction.assetID === DrawPixelAsset.ASSET_ID)
            {
                const pixel = codec.decode<DrawPixelPayload>(drawPixelSchema, transaction.asset);
                this._channel.publish("canvas:pixelChangeCommitted", {
                    address: transaction.senderAddress.toString("hex"),
                    transactionId: transaction.id.toString("hex"),
                    blockHeight: block.header.height,
                    pixel: pixel,
                });
            }
        }

        for (const canvasId of active.canvasIds)
        {
            const canvasBuffer = await stateStore.chain.get(serialiseCanvasId(canvasId));
            assert(canvasBuffer !== undefined);
            const canvas = codec.decode<CanvasPayload>(canvasSchema, canvasBuffer);
            assert(canvas.state === CanvasState.ACTIVE);

            // Automatically ending canvas after passing end time
            if (canvas.endBlockHeight <= block.header.height)
            {
                canvas.state = CanvasState.COMPLETE;
                await stateStore.chain.set(serialiseCanvasId(canvasId), codec.encode(canvasSchema, canvas));
                complete.canvasIds.push(canvasId);
                this._channel.publish("canvas:completed", { canvasId });
                continue;
            }

            activeToCommit.push(canvasId);
        }

        await stateStore.chain.set("canvas:pending", codec.encode(pendingSchema, { canvasIds: pendingToCommit }));
        await stateStore.chain.set("canvas:active", codec.encode(activeSchema, { canvasIds: activeToCommit }));
        await stateStore.chain.set("canvas:complete", codec.encode(completeSchema, complete));
    }

    public async beforeTransactionApply(_input: TransactionApplyContext): Promise<void> {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
    }

    public async afterTransactionApply(_input: TransactionApplyContext): Promise<void> {
        if (_input.transaction.moduleID === CanvasModule.MODULE_ID &&
            _input.transaction.assetID === DrawPixelAsset.ASSET_ID)
        {
            const pixel = codec.decode<DrawPixelPayload>(drawPixelSchema, _input.transaction.asset);
            this._channel.publish("canvas:pixelChangeSubmitted", {
                address: _input.transaction.senderAddress.toString("hex"),
                transactionId: _input.transaction.id.toString("hex"),
                pixel: pixel,
            });
        }
    }

    public async afterGenesisBlockApply(_input: AfterGenesisBlockApplyContext<CanvasAccount>): Promise<void> {
        const accounts = _input.genesisBlock.header.asset.accounts;
        assert(accounts.length > 0);

        for (const account of accounts)
        {
            if (account.canvas.accountType == AccountType.Admin.toString())
            {
                await _input.stateStore.chain.set("canvas:adminAddress", codec.encode(addressSchema, { address: account.address }));
            }

            if (account.canvas.accountType == AccountType.Wallet.toString())
            {
                await _input.stateStore.chain.set("canvas:walletAddress", codec.encode(addressSchema, { address: account.address }));
            }
        }
    }

    private async getCanvas(params: { canvasId: string }): Promise<CanvasResponse|null> {
        const canvasId = serialiseCanvasId(Number(params.canvasId));
        const canvasBuffer = await this._dataAccess.getChainState(canvasId);

        if (canvasBuffer === undefined)
        {
            return null;
        }

        const canvas = codec.decode<CanvasPayload>(canvasSchema, canvasBuffer);
        return {
            ...canvas,
            ownerId: canvas.ownerId.toString("hex"),
            costPerPixel: Number(canvas.costPerPixel),
            startBlockHeight: Number(canvas.startBlockHeight),
            endBlockHeight: Number(canvas.endBlockHeight),
        };
    }

    private async getPendingCanvases(): Promise<PendingPayload> {
        const pendingBuffer = await this._dataAccess.getChainState("canvas:pending");

        if (pendingBuffer === undefined)
        {
            return { canvasIds: [] };
        }

        return codec.decode<PendingPayload>(pendingSchema, pendingBuffer);
    }

    private async getActiveCanvases(): Promise<ActivePayload> {
        const activeBuffer = await this._dataAccess.getChainState("canvas:active");

        if (activeBuffer === undefined)
        {
            return { canvasIds: [] };
        }

        return codec.decode<ActivePayload>(activeSchema, activeBuffer);
    }

    private async getCompleteCanvases(): Promise<CompletePayload> {
        const completeBuffer = await this._dataAccess.getChainState("canvas:complete");

        if (completeBuffer === undefined)
        {
            return { canvasIds: [] };
        }

        return codec.decode<CompletePayload>(completeSchema, completeBuffer);
    }

    private async getWalletAddress(): Promise<Buffer|null> {
        const walletBuffer: Buffer|undefined = await this._dataAccess.getChainState("canvas:walletAddress");

        if (walletBuffer === undefined)
        {
            return null;
        }

        const wallet = codec.decode<AddressPayload>(addressSchema, walletBuffer);
        return wallet.address;
    }

    private async getAdminAddress(): Promise<Buffer|null> {
        const adminBuffer: Buffer|undefined = await this._dataAccess.getChainState("canvas:adminAddress");

        if (adminBuffer === undefined)
        {
            return null;
        }

        const admin = codec.decode<AddressPayload>(addressSchema, adminBuffer);
        return admin.address;
    }
}
