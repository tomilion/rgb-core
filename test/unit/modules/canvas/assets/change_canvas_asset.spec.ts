import { codec, testing } from "lisk-sdk";
import { ChangeCanvasAsset } from "../../../../../src/app/modules/canvas/assets/change_canvas_asset";
import { CanvasModule } from "../../../../../src/app/modules/canvas/canvas_module";
import { canvasSchema, CanvasState, ChangeCanvasPayload } from "../../../../../src/app/modules/canvas/schemas";
import { numberBetween, randomCanvas, randomAddress } from "../../../../utils/random_generator";

describe("ChangeCanvasAsset", () => {
    let mockUser: Buffer;
    let mockAsset: ChangeCanvasPayload;
    let testClass: ChangeCanvasAsset;

    beforeEach(() => {
        mockUser = randomAddress();
        mockAsset = {
            canvasId: numberBetween(0, 4294967295),
            costPerPixel: BigInt(numberBetween(0, 4294967295)),
            startBlockHeight: BigInt(numberBetween(0, 1000000)),
            endBlockHeight: BigInt(numberBetween(1000001, 2000000)),
            width: numberBetween(0, 10000),
            height: numberBetween(0, 10000),
            timeBetweenDraws: numberBetween(0, 4294967295),
            colourPalette: new Uint8Array(48),
        };
        testClass = new ChangeCanvasAsset();
    });

    describe("constructor", () => {
        it("should have valid id", () => {
            expect(testClass.id).toEqual(1);
        });

        it("should have valid name", () => {
            expect(testClass.name).toEqual("changeCanvas");
        });

        it("should have valid schema", () => {
            expect(testClass.schema).toMatchSnapshot();
        });
    });

    describe("validate", () => {
        it("should pass with valid args", () => {
            const context = testing.createValidateAssetContext({
                asset: mockAsset,
                transaction: { senderAddress: mockUser } as any,
            });
            testClass.validate(context);
        });

        it("should throw width below 0", () => {
            const context = testing.createValidateAssetContext({
                asset: { canvasId: mockAsset.canvasId, width: -1 },
                transaction: { senderAddress: mockUser } as any,
            });
            expect(() => testClass.validate(context)).toThrow("Width invalid");
        });

        it("should throw width above 10000", () => {
            const context = testing.createValidateAssetContext({
                asset: { canvasId: mockAsset.canvasId, width: 10001 },
                transaction: { senderAddress: mockUser } as any,
            });
            expect(() => testClass.validate(context)).toThrow("Width invalid");
        });

        it("should throw height below 0", () => {
            const context = testing.createValidateAssetContext({
                asset: { canvasId: mockAsset.canvasId, height: -1 },
                transaction: { senderAddress: mockUser } as any,
            });
            expect(() => testClass.validate(context)).toThrow("Height invalid");
        });

        it("should throw height above 10000", () => {
            const context = testing.createValidateAssetContext({
                asset: { canvasId: mockAsset.canvasId, height: 10001 },
                transaction: { senderAddress: mockUser } as any,
            });
            expect(() => testClass.validate(context)).toThrow("Height invalid");
        });

        it("should throw cost per pixel below 0", () => {
            const context = testing.createValidateAssetContext({
                asset: { canvasId: mockAsset.canvasId, costPerPixel: BigInt(-1) },
                transaction: { senderAddress: mockUser } as any,
            });
            expect(() => testClass.validate(context)).toThrow("Cost per pixel invalid");
        });

        it("should throw start block height in past", () => {
            const context = testing.createValidateAssetContext({
                asset: { canvasId: mockAsset.canvasId, startBlockHeight: BigInt(10000) },
                transaction: { senderAddress: mockUser } as any,
            });
            context.header = { height: 10001 } as any;
            expect(() => testClass.validate(context)).toThrow("Start block height cannot be in the past");
        });

        it("should throw end block height in past", () => {
            const context = testing.createValidateAssetContext({
                asset: { canvasId: mockAsset.canvasId, endBlockHeight: BigInt(10000) },
                transaction: { senderAddress: mockUser } as any,
            });
            context.header = { height: 10001 } as any;
            expect(() => testClass.validate(context)).toThrow("End block height cannot be in the past");
        });

        it("should throw start block height above end block height", () => {
            const context = testing.createValidateAssetContext({
                asset: { ...mockAsset, startBlockHeight: BigInt(10001), endBlockHeight: BigInt(10000) },
                transaction: { senderAddress: mockUser } as any,
            });
            expect(() => testClass.validate(context)).toThrow("End block height must be greater than start block height");
        });

        it("should throw invalid colour palette size", () => {
            const context = testing.createValidateAssetContext({
                asset: { ...mockAsset, colourPalette: new Uint8Array(47) },
                transaction: { } as any,
            });
            expect(() => testClass.validate(context)).toThrow("Colour palette invalid");
        });
    });

    describe("apply", () => {
        let account: any;
        let chain: any;

        beforeEach(() => {
            account = testing.fixtures.createDefaultAccount([CanvasModule]);
            chain = {};
        });

        it("should save canvas to state store", async () => {
            const currentAsset = randomCanvas({ ownerId: account.address, state: CanvasState.PENDING });
            chain[`canvas-${mockAsset.canvasId}`] = codec.encode(canvasSchema, currentAsset);
            const stateStore = new testing.mocks.StateStoreMock({
                accounts: [account],
                chain
            });
            const context = testing.createApplyAssetContext({
                stateStore,
                asset: mockAsset,
                transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
            });
            jest.spyOn(stateStore.chain, "set");

            await testClass.apply(context);

            expect(stateStore.chain.set).toHaveBeenCalledWith(
                `canvas-${mockAsset.canvasId}`,
                codec.encode(canvasSchema, {
                    ownerId: account.address,
                    costPerPixel: mockAsset.costPerPixel,
                    startBlockHeight: mockAsset.startBlockHeight,
                    endBlockHeight: mockAsset.endBlockHeight,
                    width: mockAsset.width,
                    height: mockAsset.height,
                    timeBetweenDraws: mockAsset.timeBetweenDraws,
                    colourPalette: mockAsset.colourPalette,
                    state: CanvasState.PENDING,
                })
            );
        });

        it("should throw canvas does not exist", async () => {
            const stateStore = new testing.mocks.StateStoreMock({
                accounts: [account]
            });
            const context = testing.createApplyAssetContext({
                stateStore,
                asset: mockAsset,
                transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
            });

            await expect(testClass.apply(context)).rejects.toThrow("Canvas does not exist");
        });

        it("should throw for invalid user", async () => {
            chain[`canvas-${mockAsset.canvasId}`] = codec.encode(canvasSchema, randomCanvas({
                ownerId: account.address,
                startBlockHeight: mockAsset.startBlockHeight,
                state: CanvasState.PENDING,
            }));
            const stateStore = new testing.mocks.StateStoreMock({
                accounts: [account],
                chain
            });
            const context = testing.createApplyAssetContext({
                stateStore,
                asset: { canvasId: mockAsset.canvasId },
                transaction: { senderAddress: randomAddress(), nonce: BigInt(1) } as any,
            });

            await expect(testClass.apply(context)).rejects.toThrow("User invalid");
        });

        it("should throw start block height above end block height", async () => {
            chain[`canvas-${mockAsset.canvasId}`] = codec.encode(canvasSchema, randomCanvas({
                ownerId: account.address,
                startBlockHeight: mockAsset.startBlockHeight,
                state: CanvasState.PENDING,
            }));
            const stateStore = new testing.mocks.StateStoreMock({
                accounts: [account],
                chain
            });
            const context = testing.createApplyAssetContext({
                stateStore,
                asset: { canvasId: mockAsset.canvasId, endBlockHeight: (mockAsset.startBlockHeight ?? BigInt(Number.MAX_SAFE_INTEGER)) - BigInt(1) },
                transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
            });

            await expect(testClass.apply(context)).rejects.toThrow("End block height must be greater than start block height");
        });

        it("should throw canvas already completed", async () => {
            chain[`canvas-${mockAsset.canvasId}`] = codec.encode(canvasSchema, randomCanvas({
                ownerId: account.address,
                state: CanvasState.COMPLETE,
            }));
            const stateStore = new testing.mocks.StateStoreMock({
                accounts: [account],
                chain
            });
            const context = testing.createApplyAssetContext({
                stateStore,
                asset: { canvasId: mockAsset.canvasId },
                transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
            });

            await expect(testClass.apply(context)).rejects.toThrow("Canvas already completed");
        });

        it("should throw canvas already started modify start block height", async () => {
            chain[`canvas-${mockAsset.canvasId}`] = codec.encode(canvasSchema, randomCanvas({
                ownerId: account.address,
                state: CanvasState.ACTIVE,
            }));
            const stateStore = new testing.mocks.StateStoreMock({
                accounts: [account],
                chain
            });
            const context = testing.createApplyAssetContext({
                stateStore,
                asset: { canvasId: mockAsset.canvasId, startBlockHeight: BigInt(1) },
                transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
            });

            await expect(testClass.apply(context)).rejects.toThrow("Cannot modify start block height on active canvas");
        });

        it("should throw canvas already started modify end block height", async () => {
            chain[`canvas-${mockAsset.canvasId}`] = codec.encode(canvasSchema, randomCanvas({
                ownerId: account.address,
                state: CanvasState.ACTIVE,
            }));
            const stateStore = new testing.mocks.StateStoreMock({
                accounts: [account],
                chain
            });
            const context = testing.createApplyAssetContext({
                stateStore,
                asset: { canvasId: mockAsset.canvasId, endBlockHeight: BigInt(1) },
                transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
            });

            await expect(testClass.apply(context)).rejects.toThrow("Cannot modify end block height on active canvas");
        });

        it("should throw start block height above end block height (existing start block height)", async () => {
            chain[`canvas-${mockAsset.canvasId}`] = codec.encode(canvasSchema, randomCanvas({
                ownerId: account.address,
                startBlockHeight: BigInt(100),
                state: CanvasState.PENDING,
            }));
            const stateStore = new testing.mocks.StateStoreMock({
                accounts: [account],
                chain
            });
            const context = testing.createApplyAssetContext({
                stateStore,
                asset: { canvasId: mockAsset.canvasId, endBlockHeight: BigInt(99) },
                transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
            });

            await expect(testClass.apply(context)).rejects.toThrow("End block height must be greater than start block height");
        });

        it("should throw start block height above end block height (existing end block height)", async () => {
            chain[`canvas-${mockAsset.canvasId}`] = codec.encode(canvasSchema, randomCanvas({
                ownerId: account.address,
                endBlockHeight: BigInt(100),
                state: CanvasState.PENDING,
            }));
            const stateStore = new testing.mocks.StateStoreMock({
                accounts: [account],
                chain
            });
            const context = testing.createApplyAssetContext({
                stateStore,
                asset: { canvasId: mockAsset.canvasId, startBlockHeight: BigInt(101) },
                transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
            });

            await expect(testClass.apply(context)).rejects.toThrow("End block height must be greater than start block height");
        });
    });
});
