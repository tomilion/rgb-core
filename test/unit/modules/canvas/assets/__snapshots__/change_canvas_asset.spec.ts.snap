// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeCanvasAsset constructor should have valid schema 1`] = `
Object {
  "$id": "canvas/changeCanvas-asset",
  "properties": Object {
    "canvasId": Object {
      "dataType": "uint32",
      "fieldNumber": 1,
    },
    "costPerPixel": Object {
      "dataType": "uint64",
      "fieldNumber": 2,
    },
    "endBlockHeight": Object {
      "dataType": "uint64",
      "fieldNumber": 4,
    },
    "height": Object {
      "dataType": "uint32",
      "fieldNumber": 6,
    },
    "startBlockHeight": Object {
      "dataType": "uint64",
      "fieldNumber": 3,
    },
    "timeBetweenDraws": Object {
      "dataType": "uint32",
      "fieldNumber": 7,
    },
    "width": Object {
      "dataType": "uint32",
      "fieldNumber": 5,
    },
  },
  "required": Array [
    "canvasId",
  ],
  "title": "ChangeCanvasAsset transaction asset for canvas module",
  "type": "object",
}
`;
